### ISSUE TITLE
bnorm(Batch Normalization) layer for testing

### ISSUE THREAD

**Issue by jonghyuk65** [vlfeat/matconvnet#263]
Hi everyone.

Recently, I was working on using batch normalization layer.
It works well in training phase.
But in testing phase, I realized that I need to compute overall(through all training or validation set) mean and variance of filter responses.

Issue 193 (https://github.com/vlfeat/matconvnet/issues/193) talked about same problem.
Is there any update about this issue?

Or, can anyone suggest simple way to force overall mean and variance to bnorm layer?
I can't find easy way, except modifying c++ codes in matconvnet.

Thanks!

**Comment by vedaldi** [comment]
Hi,

yes, definitely we should make this process easier and we are thinking what would be the best way to do so in term of the interface (BTW, the “batch-normalization off” version of the block could be written in MATLAB directly without sensible speed hit I think).

However, there *is* a relatively simple fix that is also more efficient. After training, you should be able to incorporate mean and variance in the following bank of linear filters and remove bnorm.

**Comment by kjw0612** [comment]
Don't we also need to keep track of statistics during training?

**Comment by vedaldi** [comment]
Hi, during training statistics are computed afresh for each batch. During testing, you would first record a set of statistics from many images, and then freeze them. 

This is the “annoying” part that requires two or even three different code paths, one for training, one for testing, and one for collecting the final statistics, but it is also necessary if you want to run the network on single images (from which it would not be possible to estimate any sensible statistics).

It may be possible to accumulate the required statistics during training. In particular, in the DAG version you could achieve this by internal variables to the bnorm block. Then, it would be possible to switch bnorm to a ‘test’ mode to use these stats. However, this would slow down training somewhat and I am a bit hesitant to go this way.

Note that batch normalisation introduces a lot of noise during training due to the fluctuations in the statistics; but apparently this is a good thing (it replaces dropout).

**Comment by kjw0612** [comment]
Thanks for explanation. Hope someone come up with a smart solution!

**Comment by jonghyuk65** [comment]
Thanks to your helps, I could understand the current issues and temporal solution for me.
Hope you find clear solution.

Thanks!

**Comment by AruniRC** [comment]
Hi,
Just to add something small: there happens to be a getImageStats() function in the cnn_imagenet.m that is used to calculate the average image if the network normalization parameters does not already have an average image defined. 
Thus, the basic setup for looping over all the batches is already present. It seems reasonable to compute not just the average image, but the statistics of the features by applying the CNN for each image we are reading in. Not particularly efficient, but should work.

**Comment by nicjac** [comment]
I have also been experimenting with batch normalization and I am slightly confused as to what should be done at test time. Similar to others, I get the same identical output from the network regardless of the image supplied. I know why this is, but I am not sure what to do about it. 

Shouldn't the image statistics be automatically computed at inference when given a 4D tensor (x,y,1,n) where n is the number of grayscale images in a batch? If not, how does it work during the evaluation phase following each training epoch? How are the statistics computed then? Combing through the code did not help me much here.

**Comment by etienne87** [comment]
@vedaldi : i'm trying this very simple stuff with caffe... https://groups.google.com/forum/#!topic/caffe-users/Cz-w0xoOQfo  could someone point to me what i'm doing wrong?

**Comment by lenck** [comment]
Ok, this has actually been solved in some of the previous MatConvNet releases. Now BNorm has multiple outputs where the mean and variance output is taken as a parameter which is being learnt in `cnn_train` using a [simple averaging](https://github.com/vlfeat/matconvnet/blob/master/examples/cnn_train_dag.m#L290) (instead of SGD with momentum).
As those are parameters, they are then provided to the BatchNorm during the test mode, instead of being computed on the fly on the provided data (during training in the `normal` mode), see e.g. [this line](https://github.com/vlfeat/matconvnet/blob/master/matlab/%2Bdagnn/BatchNorm.m#L10).
Also, note that when the network is deployed, the BatchNormalization can be removed entirely as the moments [can be merged to the weights of the previous convolution layer](https://github.com/vlfeat/matconvnet/blob/master/examples/imagenet/cnn_imagenet_deploy.m#L176). 
@etienne87 Considering that you have solved your caffe issue, I hope it is ok that I will close this one.

**Comment by omair-kg** [comment]
@vedaldi 
Hi Andrea,

You mentioned above 
 

> Note that batch normalisation introduces a lot of noise during training due to the fluctuations in the statistics; but apparently this is a good thing (it replaces dropout)

Could you elaborate what you mean by replaces dropout. Thanks in anticipation.

BR
Omair

**Comment by vedaldi** [comment]
Hi, dropout inject noise in the network for the purpose of regularisation during training. Batch normalisation has (implicitly) a similar effect, although the type of noise is quite different.

**Comment by omair-kg** [comment]
Thanks
